### DISCUSSION TITLE
How to accept optional fields with empty or NULL value in adonis v5

### DISCUSSION THREAD

**Discussion by zmaglica** [adonisjs/core#1388] [Help]
Hi. Let's have a case where the end-user wants to update his profile and he wants to remove his phone number (for example). Since it is an update form, all fields are optional. The problem is when I send a null, validation ignores it and `request.validate` function doesn't return the "phone" field that is null in this case.

When I try it using Insomnia (or postman) with phone parameter equals NULL, validator gets it as a string and saves it as a  string

Does anyone know how to achieve it?

**Comment by fnoquiq** [comment]
You need to add that `phone` is optional.

```ts
public schema = schema.create({
  field1: schema.string(),
  field2: schema.string.opcional(),
})
```

```ts
{
  "field1": "something",
  "field2": null // Undefined
}

// or

{
  "field1": "something"
} // No need to send the optional field in the body of the request
```

```ts
const { field1, field2 } = await request.validate(validator)

// In both cases above, this value will be undefined
console.log(field2) // Undefined
```

**Reply by zmaglica** [reply]
yeah... but I need null. I want to update the phone to null value

**Reply by thetutlage** [reply]
Couple of things here.

First, you want to skip the required validation. That can be done by chaining the `.optional()` method and using `existsStrict` option.

```ts
  const body = await request.validate({
    schema: schema.create({
      phone: schema.string.optional(),
    }),
    existsStrict: true,
  })
```

But now, the validator will complain that `null` is not a string. It means, you need to send an empty string from the form and not null.

```json
{
  "phone": ""
}
``` 

And inside your controller, you can have a check.

```ts
if (body.phone === '') {
  // set to null
}
```

**Reply by zmaglica** [reply]
Thank you for your comment. How can I set `existsStrict: true, `when using the validator like this: `const data = await ctx.request.validate(UpdateAccountUserValidator)` ?

**Reply by thetutlage** [reply]
You can set it as a public property on the class.

```ts
class UpdateAccountUserValidator {
   public existsStrict = true
}
```

**Reply by zmaglica** [reply]
@thetutlage I tried the solution you provided and I phone is not there when the validator is finished with validation.

here is the schema/validator file
```
import {HttpContextContract} from '@ioc:Adonis/Core/HttpContext'
import {schema, rules} from '@ioc:Adonis/Core/Validator'
import AccountUser from "App/Domains/Account/Models/AccountUser";


export default class UpdateAccountUserValidator {

  constructor(private ctx: HttpContextContract) {
  }

  public schema = schema.create({
    phone: schema.string.optional(),
  })
  public existsStrict = true
  public cacheKey = this.ctx.routeKey
  public messages = {}
}
```

And inside a controller:

```
    const data = await ctx.request.validate(UpdateAccountUserValidator)
    console.log(data)
```

Console.log returns an empty object

**Reply by thetutlage** [reply]
What is the data your are submitting?

**Reply by zmaglica** [reply]
![image](https://user-images.githubusercontent.com/15913573/89512876-201c8500-d7d4-11ea-8f86-5f3b58c2ee8a.png)

Here is the screenshot of insomnia

**Reply by thetutlage** [reply]
I just said, you should send an empty string. Read my comment carefully

**Reply by zmaglica** [reply]
Ok so null is not an option here? I need to send something else or create my own validation rule? Because the frontend sends it as null because empty string is parsed as a string and quotes are saved in the database.

anyway, I will parse it differently without relying on the validator.

thank you for your help.

**Reply by fnoquiq** [reply]
Or 

```ts
if ( !phone ) {
  phone = null
}

**Reply by zmaglica** [reply]
@fnoquiq I am planning to do something similar. If "phone" parameter exists in the request object and it is empty, null I will cast it to null.

**Reply by thetutlage** [reply]
The data type of phone is a string and `null` is not a string. So whether you omit the field all together or you send an empty string. There is no 3rd option

**Reply by fnoquiq** [reply]
Maybe something like this works too. 

```ts
user.merge({ phone: phone ?? null })
await user.save()
```

**Note:** [*not tested*]

**Reply by zmaglica** [reply]
@fnoquiq it will work in case that front-end always sends phone parameter. In cases where phone parameter is missing (when you don't want to change your phone), your example will change phone value to null.

Thank you for example.
